transcom/mymove
# Try to fix GHA spurious notifications

## Summary

As of 2023-09-11, we get spurious notifications from GitHub because the analyze bundle action is triggered on both `pull_request` and `pull_request_target` (that's my fault). After reading more about GHA, it seems to me that `pull_request` is better for MilMove since all our requests come from the same repo and not a fork and it's safer if external forks were ever a thing.

In addition, I've added support for canceling in progress jobs to save some processing (and money). Unfortunately [GitHub doesn't support suppressing cancelation notifications](https://github.com/orgs/community/discussions/13015) and so this change may actually increase notifications. I still think the reduction in redundant processing is worth it.

I added caching of the bundle stats for the analyze bundle job. For the base, that's almost always `main` and we should get decent hit rates for that on PRs, saving even more processing. For the PR, if somehow the PR runs again with the same commit sha, the cache should help. That's probably rare, but any hit helps.

I added caching for yarn to hopefully speed up dependency installation.

Because this PR switches from `pull_request_target`, the target runs for the bundle diff will still run. That means you will see two runs in the `Actions` tab, one for the branch and one without a branch. The 2nd one will go away once this is merged.

## Verification Steps for the Author

I've watched the Actions and it seems they work fine.

See [this run](https://github.com/transcom/mymove/actions/runs/6148331267/job/16681811796) for a run without a bundle cache hit and [this run](https://github.com/transcom/mymove/actions/runs/6149110176/job/16684301293) for a run with a bundle cache hit.

See [this run](https://github.com/transcom/mymove/actions/runs/6148611972/job/16682686126) for a run without a dependency caching hit and [this run](https://github.com/transcom/mymove/actions/runs/6148926517/job/16683729481) for a run with a dependency cache hit.

## Verification Steps for Reviewers

All of the changes are for GHA, so there's not a really great way to test locally.
